source/common/runtime/runtime_impl.h

```diff
@@ -31,7 +31,10 @@ namespace Runtime {
 class RandomGeneratorImpl : public RandomGenerator {
 public:
   // Runtime::RandomGenerator
-  uint64_t random() override { return threadLocalGenerator()(); }
+  uint64_t random() override {
+    auto& gen = threadLocalGenerator();
+    return (gen() << 48) | gen();
```

Comment:
This is not free and I'm hugely in favor of calling this twice every time we want a random number. Can you explain the underlying issue more?

Reply:
The hashing load balancer distributes its N hosts over 1024 buckets, then for each request, selects the nearest bucket to the hash value for the request. If no hash policy matches, the "hash" is randomly generated. A randomly-generated hash, though, has 0's as the top 16 bits. Since the "nearest" bucket is the one that has the smallest value greater than the request's has, the result is that the lowest-valued bucket is always selected, so only one host is ever routed to.

I agree that this is not free, and is probably not the best way to fix this particular degenerate case. It is, however, consistent with the declared API.

Reply:
If perf is a concern what do you think of splitting/renaming?    If we had a random48() and random64() it'd solve the perf problem and increase clarity

Reply:
If you stick with something similar to this, I think you really want to XOR the upper and lower parts, instead of OR.  Otherwise I think you'll bias towards 1 instead of 0 for bits 16-48.

Reply:
Renaming is an option. Not sure if it's worth doing a bulk rename through the entire code base, but could add a new random64() function or something. 

Another option is to switch the underlying pseudo random generator to one that spits out 64 bits. I'm not super familiar with the algorithms but we could look into this.

Another option is to special case the ring hash LB and just swap the returned bytes (just flip each 32-bit quad). That requires knowledge of the underlying algorithm but is then localized to the actual problem site.

Reply:
Was there a reason to select `std::ranlux48` in the first place? Or was this arbitrary?

Reply:
I think I did some research and this was recommended as a good balance of randomness and speed.

Reply:
Regardless of what is done elsewhere, we should either rename/add or change the documentation for random() because it's misleading.

Reply:
Yup agreed. Can you maybe look and see if there is a different algorithm in std:: that we can use to get 64-bits. Also, there might be something in boringssl. cc @PiotrSikora

Reply:
I've added generic `RandomGeneratorImpl::random_bytes()` as part [work on UUID]( https://github.com/envoyproxy/envoy/pull/1465/commits/1d76e26c9857764027436b38eab72cca12a2cbc4), but it was later simplified and inlined into `RandomGeneratorImpl::uuid()`.

We might want to revisit that code (if it's not slower than `ranlux48`) instead of rolling our own pseudo-randomness.